src/spawn.rs

```diff
@@ -1,116 +1,241 @@
+    pub fn burst(count: Value<f32>, period: Value<f32>) -> Self {
+        Self::new(count, 0.0.into(), period)
+    }
+
+    /// Resets and activates the spawner.
+    /// Use this, for example, to immediately spawn some particles.
+    pub fn reset(&mut self) {
+        self.time = 0.;
+        self.limit = 0.;
+        self.spawn = 0.5;
+        self.active = true;
+    }
+
+    /// Sets whether the spawner is active.
+    pub fn set_active(&mut self, active: bool) {
+        self.active = active;
+    }
+
+    /// Resamples the spawn time and period.
+    fn resample(&mut self) {
+        self.curr_spawn_time = self.spawn_time.sample();
+        self.limit = self.period.sample();
+    }
+
+    pub(crate) fn tick(&mut self, mut dt: f32) -> u32 {
+        if !self.active {
+            return 0;
+        }
+
+        // The limit can be reached multiple times, so use a loop
+        loop {
+            if self.limit == 0.0 {
+                self.resample();
+                continue;
+            }
+
+            let new_time = self.time + dt;
+            if self.time <= self.curr_spawn_time {
+                self.spawn += if self.curr_spawn_time == 0.0 {
+                    self.num_particles.sample()
+                } else if new_time <= self.curr_spawn_time {
+                    self.num_particles.sample() * dt / self.curr_spawn_time
+                } else {
+                    self.num_particles.sample() * (1.0 - self.time / self.curr_spawn_time)
+                };
+            }
+
+            let old_time = self.time;
+            self.time = new_time;
+
+            if self.time >= self.limit {
+                dt -= self.limit - old_time;
+                self.time = 0.0; // dt will be added on in the next iteration
+                self.resample();
+            } else {
+                break;
+            }
+        }
+
+        let count = self.spawn.floor();
+        self.spawn -= count;
+        count as u32
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_once() {
+        let mut spawner = Spawner::once(5.0.into());
+        let count = spawner.tick(0.001);
+        assert_eq!(count, 5);
+        let count = spawner.tick(100.0);
+        assert_eq!(count, 0);
+    }
+
+    #[test]
+    fn test_once_reset() {
+        let mut spawner = Spawner::once(5.0.into());
+        spawner.tick(1.0);
+        spawner.reset();
+        let count = spawner.tick(1.0);
+        assert_eq!(count, 5);
+    }
+
+    #[test]
+    fn test_once_inactive() {
+        let mut spawner = Spawner::once(5.0.into()).with_active(false);
+        let count = spawner.tick(1.0);
+        assert_eq!(count, 0);
+        spawner.reset();
+        let count = spawner.tick(1.0);
+        assert_eq!(count, 5);
+    }
+
+    #[test]
+    fn test_rate() {
+        let mut spawner = Spawner::rate(5.0.into());
+        let count = spawner.tick(1.0);
+        assert_eq!(count, 5);
+        let count = spawner.tick(0.4);
```

Comment:
I'd be more careful and make sure even with numerical imprecision we're past the 40% mark, e.g.
```suggestion
        let count = spawner.tick(0.41);
```
Maybe that works now because of `0.4` but I don't expect that'd work exactly with other values.

Reply:
It works now because I initialized `spawn` to 0.5, which should make it stay clear of the edge case. I could make the change and then also allow the total count to be off by 1 on the high end.

Reply:
I don't understand, the spawner is spawning 5 particles per second. On first tick it goes from t=0 to t=1, spawning 5, which is already on an edge case (could be 4.99999 and `floor()` would round to 4, failing the assert). Then on second tick it goes from t=1 to t=1.4, again on an edge case spawning (5.0 * 0.4) = 2 particles. No?